File: backend/app/models/workflows.py
"""
Workflow Models - Stage 5

Task templates, onboarding tasks, escalation rules.
"""
from datetime import datetime
from typing import Optional, List
from enum import StrEnum

from beanie import Document
from pydantic import Field


class TaskTemplate(Document):
    """Tenant-configurable task templates for incidents and safety."""
    
    tenant_id: str = Field(..., index=True)
    template_name: str = Field(..., index=True)
    template_type: str = Field(..., index=True)  # "incident", "safety", "custom"
    title: str
    description: Optional[str] = None
    priority: Optional[str] = None  # "low", "medium", "high", "urgent"
    status: Optional[str] = None  # Default status when created
    is_locked: bool = Field(default=False)  # Locked workflows cannot be modified
    template_data: Optional[dict] = None  # renamed from 'metadata' (reserved)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "task_templates"
        indexes = [
            "tenant_id",
            "template_name",
            "template_type",
        ]


class OnboardingTaskSource(StrEnum):
    """Source of auto-generated onboarding task."""
    MODULE = "module"
    INDUSTRY = "industry"
    COMPLIANCE_RULE = "compliance_rule"


class OnboardingTask(Document):
    """Auto-generated onboarding tasks."""
    
    tenant_id: str = Field(..., index=True)
    task_id: Optional[str] = Field(default=None, index=True)  # Link to actual Task if created
    source: OnboardingTaskSource = Field(..., index=True)
    source_id: Optional[str] = None  # e.g., module code, industry code, rule code
    title: str
    description: Optional[str] = None
    is_required: bool = Field(default=True)  # Non-deletable required tasks
    is_completed: bool = Field(default=False)
    completed_at: Optional[datetime] = None
    due_date: Optional[datetime] = None
    assigned_to_user_id: Optional[str] = Field(default=None, index=True)
    created_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "onboarding_tasks"
        indexes = [
            "tenant_id",
            "task_id",
            "source",
            "assigned_to_user_id",
        ]


class EscalationRule(Document):
    """Escalation rules for overdue required tasks."""
    
    tenant_id: str = Field(..., index=True)
    rule_name: str
    trigger_condition: str  # "overdue_required_task", "not_completed_by_due_date"
    escalation_delay_hours: int = Field(default=24)  # Hours after due date before escalation
    notify_roles: Optional[List[str]] = None  # ["owner", "manager"]
    is_active: bool = Field(default=True)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "escalation_rules"
        indexes = [
            "tenant_id",
        ]


class EscalationEvent(Document):
    """Log of escalation events."""
    
    tenant_id: str = Field(..., index=True)
    task_id: str = Field(..., index=True)
    rule_id: Optional[str] = Field(default=None, index=True)
    escalated_at: datetime = Field(default_factory=datetime.utcnow)
    notified_user_ids: Optional[List[str]] = None
    resolved_at: Optional[datetime] = None

    class Settings:
        name = "escalation_events"
        indexes = [
            "tenant_id",
            "task_id",
            "rule_id",
        ]

File: frontend/lib/api/workflows.ts
import apiClient from "./client";

// ========== Stage 5: Task Templates ==========

export interface TaskTemplate {
  id: number;
  tenant_id: number;
  template_name: string;
  template_type: string;
  title: string;
  description: string | null;
  priority: string | null;
  status: string | null;
  is_locked: boolean;
  metadata: Record<string, any> | null;
  created_at: string;
  updated_at: string;
}

export interface TaskTemplateCreate {
  template_name: string;
  template_type: "incident" | "safety" | "custom";
  title: string;
  description?: string;
  priority?: string;
  status?: string;
  is_locked?: boolean;
  metadata?: Record<string, any>;
}

export const createTaskTemplate = async (data: TaskTemplateCreate): Promise<TaskTemplate> => {
  const response = await apiClient.post("/workflows/templates", data);
  return response.data;
};

export const getTaskTemplates = async (templateType?: string): Promise<TaskTemplate[]> => {
  const url = templateType 
    ? `/workflows/templates?template_type=${templateType}`
    : "/workflows/templates";
  const response = await apiClient.get(url);
  return response.data;
};

export const updateTaskTemplate = async (templateId: number, data: TaskTemplateCreate): Promise<TaskTemplate> => {
  const response = await apiClient.put(`/workflows/templates/${templateId}`, data);
  return response.data;
};

export const deleteTaskTemplate = async (templateId: number): Promise<void> => {
  await apiClient.delete(`/workflows/templates/${templateId}`);
};

export const createTaskFromTemplate = async (
  templateId: number,
  projectId: number,
  title?: string,
  description?: string
): Promise<{ task_id: number }> => {
  const response = await apiClient.post(`/workflows/templates/${templateId}/create-task`, {
    template_id: templateId,
    project_id: projectId,
    title,
    description,
  });
  return response.data;
};

// ========== Stage 5: Onboarding Tasks ==========

export interface OnboardingTask {
  id: number;
  tenant_id: number;
  task_id: number | null;
  source: "module" | "industry" | "compliance_rule";
  source_id: string | null;
  title: string;
  description: string | null;
  is_required: boolean;
  is_completed: boolean;
  completed_at: string | null;
  due_date: string | null;
  assigned_to_user_id: number | null;
  created_at: string;
}

export const generateOnboardingTasks = async (assignedToUserId?: number): Promise<{ tasks_generated: number; tasks: OnboardingTask[] }> => {
  const url = assignedToUserId
    ? `/workflows/onboarding-tasks/generate?assigned_to_user_id=${assignedToUserId}`
    : "/workflows/onboarding-tasks/generate";
  const response = await apiClient.post(url);
  return response.data;
};

export const getOnboardingTasks = async (): Promise<OnboardingTask[]> => {
  const response = await apiClient.get("/workflows/onboarding-tasks");
  return response.data;
};

export const createTaskFromOnboardingTask = async (
  onboardingTaskId: number,
  projectId: number
): Promise<{ task_id: number }> => {
  const response = await apiClient.post(`/workflows/onboarding-tasks/${onboardingTaskId}/create-task?project_id=${projectId}`);
  return response.data;
};

// ========== Stage 5: Escalation Rules ==========

export interface EscalationRule {
  id: number;
  tenant_id: number;
  rule_name: string;
  trigger_condition: string;
  escalation_delay_hours: number;
  notify_roles: string[] | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface EscalationRuleCreate {
  rule_name: string;
  trigger_condition: "overdue_required_task" | "not_completed_by_due_date";
  escalation_delay_hours?: number;
  notify_roles?: string[];
}

export const createEscalationRule = async (data: EscalationRuleCreate): Promise<EscalationRule> => {
  const response = await apiClient.post("/workflows/escalation-rules", data);
  return response.data;
};

export const getEscalationRules = async (): Promise<EscalationRule[]> => {
  const response = await apiClient.get("/workflows/escalation-rules");
  return response.data;
};

export const checkEscalations = async (): Promise<{ escalations_triggered: number; events: Array<{ id: number; task_id: number }> }> => {
  const response = await apiClient.post("/workflows/escalation-rules/check");
  return response.data;
};

